src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -1823,110 +1854,189 @@ private RegexNode ReducePrevent()
+                while (true)
+                {
+                    RegexNode? parent = subsequent.Next;
+                    switch (parent?.Type)
                     {
-                        case One when !RegexCharClass.CharInClass(subsequent.Ch, node.Str!):
-                        case Onelazy or Oneloop or Oneloopatomic when subsequent.M > 0 && !RegexCharClass.CharInClass(subsequent.Ch, node.Str!):
-                        case Multi when !RegexCharClass.CharInClass(subsequent.Str![0], node.Str!):
-                        case Set when !RegexCharClass.MayOverlap(node.Str!, subsequent.Str!):
-                        case Setlazy or Setloop or Setloopatomic when subsequent.M > 0 && !RegexCharClass.MayOverlap(node.Str!, subsequent.Str!):
-                        case End:
-                        case EndZ or Eol when !RegexCharClass.CharInClass('\n', node.Str!):
-                        case Boundary when node.Str == RegexCharClass.WordClass || node.Str == RegexCharClass.DigitClass:
-                        case NonBoundary when node.Str == RegexCharClass.NotWordClass || node.Str == RegexCharClass.NotDigitClass:
-                        case ECMABoundary when node.Str == RegexCharClass.ECMAWordClass || node.Str == RegexCharClass.ECMADigitClass:
-                        case NonECMABoundary when node.Str == RegexCharClass.NotECMAWordClass || node.Str == RegexCharClass.NotDigitClass:
+                        case Atomic:
+                        case Alternate:
+                        case Capture:
+                            subsequent = parent;
+                            continue;
+
+                        case Concatenate:
+                            var peers = (List<RegexNode>)parent.Children!;
+                            int currentIndex = peers.IndexOf(subsequent);
+                            Debug.Assert(currentIndex >= 0, "Node should have been in its parent's child list");
+                            if (currentIndex + 1 == peers.Count)
+                            {
+                                subsequent = parent;
+                                continue;
+                            }
+                            else
+                            {
+                                subsequent = peers[currentIndex + 1];
+                                break;
+                            }
+
+                        case null:
+                            // If we hit the root, we're at the end of the expression, at which point nothing could backtrack
+                            // in and we can declare success.
                             return true;
+
+                        default:
+                            // Anything else, we don't know what to do, so we have to assume it could conflict with the loop.
+                            return false;
                     }
+
                     break;
+                }
```

Comment:
seems like there should be a return false here. I must be misreading.

Reply:
If we get here, we need to loop around the outermost loop again with the new subsequent we just found.